docs/design-documents/20231117-better-processors.md

```diff
@@ -0,0 +1,889 @@
+interface to control the processor lifecycle:
+
+```go
+// Processor receives records, manipulates them and returns back the processed
+// records.
+type Processor interface {
+	// Specification contains the metadata of this processor like name, version,
+	// description and a list of parameters expected in the configuration.
+	Specification() Specification
+
+	// Configure is the first function to be called in a processor. It provides the
+	// processor with the configuration that needs to be validated and stored.
+	// In case the configuration is not valid it should return an error.
+	// Configure should not open connections or any other resources. It should solely
+	// focus on parsing and validating the configuration itself.
+	Configure(context.Context, map[string]string) error
+
+	// Open is called after Configure to signal the processor it can prepare to
+	// start writing records. If needed, the processor should open connections and
+	// start background jobs in this function.
+	Open(context.Context) error
+
+	// Process takes a number of records and processes them right away.
+	// It should return a slice of ProcessedRecord that matches the length of
+	// the input slice. If an error occurred while processing a specific record
+	// it should be reflected in the ProcessedRecord with the same index as the
+	// input record that caused the error.
+	// Process should be idempotent, as it may be called multiple times with the
+	// same records (e.g. after a restart when records were not flushed).
+	Process(context.Context, []Record) ([]ProcessedRecord)
+
+	// Teardown signals to the processor that the pipeline is shutting down and
+	// there will be no more calls to any other function. After Teardown returns,
+	// the processor will be discarded.
+	Teardown(context.Context) error
+}
+```
+
+Notice that the processor returns its own specification in the first method. This
+is required so that Conduit can build a registry of all available processors
+and give the user more information about them (see
+[Listing processor plugins](#listing-processor-plugins)).
+
+A few notes about the `Process` method:
+
+- Even though we are not adding support for processing multiple records at the
+  same time, we propose to design the interface in a way that will let us add
+  that functionality in the future without breaking changes. This means that the
+  function will for now be called only with a single record.
+- Notice that the return value is a slice of type `ProcessedRecord`. The purpose
+  behind this is to be able to figure out what happened to each input record. This
+  way we can associate each input record to an output record and know what happened
+  to it (was it filtered out, split, merged, or only transformed). This distinction
+  is important because of acknowledgments (e.g. if a record is split into multiple
+  records, Conduit will acknowledge the original record only once all split records
+  are acknowledged).
```

Comment:
Given that a ProcessedRecord will somehow be associated with the original record, we should somehow call that out and/or provide functions in the Processors SDK that help do that. It's not really part of the design, but calling it out since I believe it will be important (because not properly associating processed records with input records can lead to unexpected behavior).

Reply:
I imagined the association to be done by the index of the record in the slice. So if you get 2 records `[rec1, rec2]` and want to filter out the first one, you should return a slice with `[filtered, rec2]` (we need to still see how we implement this `filtered` entity). This allows us to support splitting and merging records in the future and associating the merged/split record with a specific input record.

Reply:
So in the case of splitting records, we'd get a ProcessedRecord, which actually has a slice of records in it? And for a merged record in the example you mentioned, we'd get something like `[rec1, merged]`?

Reply:
(Btw, I don't think it's a must to have this discussed and nailed right away, we can decide on the details in the implementation.)

Reply:
Yes, so splitting a record would mean create a `ProcessedRecord` that itself contains multiple records so we can associate them all with the original record.

For merging I'm not exactly sure. When merging it would be nice if we could associate the merged record with all records that were merged. Once that merged record gets acked, we forward the ack for all original records that were merged into this rec. So, to get back to the example, if you get records `[rec1, rec2]` and you merge them, the result should probably be two merged records `[merged, merged]` that are somehow associated with each other, so Conduit knows it's the same record for both original records.

As you said, the details should be decided once we actually get to merging / splitting of records.